### ISSUE TITLE
[Feature Request] Relax broadcasting rules for binray opertators.

### ISSUE THREAD

**Issue by marty1885** [tenstorrent/tt-metal#14502]
**Is your feature request related to a problem? Please describe.**
Currently binary TTNN operators follows the NumPy broadcasting rules. i.e. only dimensions of (implied) 1 can be broadcasted. Eg the following works

```plaintext
tensor a:   1 x    1 x   1 x    16
tensor b:   1 x    1 x   16 x   16
+)
------------------------------------
tensor c:   1 x    1 x   16 x   16
```

But this is disallowed:

```plaintext
tensor a:   1 x    2 x   1 x    16
tensor b:   1 x    4 x   16 x   16
+)
------------------------------------
ERROR: Dimension trying to broadcast is not equal to 1
```

However this kind of broadcasting is common in GGML as a part of masking in softmax and parts of non-transformer LLMs. GGML allows a very loose broadcasting rule for binary operations. Simply requesting every dimension to be a modulo of it's counterpart.

```c
// check if t1 can be represented as a repeatition of t0
bool ggml_can_repeat(const struct ggml_tensor * t0, const struct ggml_tensor * t1) {
    static_assert(GGML_MAX_DIMS == 4, "GGML_MAX_DIMS is not 4 - update this function");

    return ggml_is_empty(t0) ? ggml_is_empty(t1) :
        (t1->ne[0]%t0->ne[0] == 0) &&
        (t1->ne[1]%t0->ne[1] == 0) &&
        (t1->ne[2]%t0->ne[2] == 0) &&
        (t1->ne[3]%t0->ne[3] == 0);
}
```

**Describe the solution you'd like**
Relax the broadcasting rules for binary operator (+-*) in TTNN

**Describe alternatives you've considered**

**Additional context**

**Comment by marty1885** [comment]
Update on my side. This feature is needed to also properly support Softmax on GGML. Currently his lack of loose broadcasting is leading to ~30% of all operations in TinyLLaMA be moved to the CPU.

**Comment by eyonland** [comment]
We are actively working on broadcasting.  Eta is November 22, 2024.

**Comment by zzigler-tt** [comment]
@patrickroberts @cmaryanTT how is this coming along? Is this a bounty candidate?
@adifrancescoTT fyi

**Comment by cmaryanTT** [comment]
@zzigler-tt - I already vetoed this one for a bounty for the immediate future.

**Comment by minaliuTT** [comment]
@patrickroberts Just checking, is this issue still valid, or can it be closed?

**Comment by marty1885** [comment]
So now I am in the company this is what I know

1. The FPU can move Dst -> Src{A,B}
2. There are some but near non existence (exist for reduce) API support for FPU to directly operate off Src.
3. This introduces weird manual latency management which is fun to deal with

**Comment by cmaryanTT** [comment]
@dchenTT - can this be closed now

**Comment by dchenTT** [comment]
@cmaryanTT Yes, we can. We only support broadcast from 1. For the case in description, user can use ttnn.repeat
